app/actions/TaxYearAction.scala

```diff
@@ -16,51 +16,50 @@
 
 package actions
 
-import common.SessionValues
 import common.SessionValues._
 import config.AppConfig
 import models.AuthorisationRequest
 import play.api.Logger
-import play.api.i18n.{I18nSupport, MessagesApi}
 import play.api.mvc.Results.Redirect
 import play.api.mvc._
 
 import javax.inject.Inject
 import scala.concurrent.{ExecutionContext, Future}
 
-class TaxYearAction @Inject()(taxYear: Int, missingTaxYearReset: Boolean)
-                             (implicit val appConfig: AppConfig, val messagesApi: MessagesApi)
-  extends ActionRefiner[AuthorisationRequest, AuthorisationRequest] with I18nSupport {
+class TaxYearAction @Inject()(taxYear: Int)
+                             (implicit val appConfig: AppConfig)
+  extends ActionRefiner[AuthorisationRequest, AuthorisationRequest] {
 
   implicit val executionContext: ExecutionContext = ExecutionContext.global
   lazy val logger: Logger = Logger.apply(this.getClass)
 
   override def refine[A](request: AuthorisationRequest[A]): Future[Either[Result, AuthorisationRequest[A]]] = {
     implicit val implicitUser: AuthorisationRequest[A] = request
 
-    Future.successful(
-      if (!appConfig.taxYearErrorFeature || taxYear == appConfig.defaultTaxYear) {
-        val sameTaxYear = request.session.get(TAX_YEAR).exists(_.toInt == taxYear)
+    lazy val overviewRedirect = Left(Redirect(appConfig.incomeTaxSubmissionOverviewUrl(taxYear)).addingToSession(TAX_YEAR -> taxYear.toString))
 
-        if (sameTaxYear || !missingTaxYearReset) {
+    def taxYearListCheck(validTaxYears: Seq[Int]): Either[Result, AuthorisationRequest[A]] = {
+      if (!appConfig.taxYearErrorFeature || validTaxYears.contains(taxYear)) {
+        val sameTaxYear = request.session.get(TAX_YEAR).exists(_.toInt == taxYear)
+        if (sameTaxYear) {
           Right(request)
         } else {
           logger.info("[TaxYearAction][refine] Tax year provided is different than that in session. Redirecting to overview.")
-          Left(
-            Redirect(appConfig.incomeTaxSubmissionOverviewUrl(taxYear))
-              .addingToSession(TAX_YEAR -> taxYear.toString)
-          )
+          overviewRedirect
         }
       } else {
         logger.info(s"Invalid tax year, adding default tax year to session")
-        Left(Redirect(controllers.errors.routes.TaxYearErrorController.show())
-          .addingToSession(SessionValues.TAX_YEAR -> appConfig.defaultTaxYear.toString)(request))
+        overviewRedirect
       }
-    )
+    }
+
+    val validTaxYears = request.session.get(VALID_TAX_YEARS)
+
+    Future.successful( if (validTaxYears.isEmpty) overviewRedirect else taxYearListCheck(validTaxYears.get.split(",").toSeq.map(_.toInt)) )
```

Comment:
Same here, We should probably, redirect to the Start page rather than the overview page, just to save on unnecessary redirects. Would be good to have a logger here as well